test/data_file_test.cpp

```diff
@@ -144,12 +145,49 @@ TEST(DataFileTest, IterateOverRecords)
 	};
 
 	unsigned row = 0;
-	for (FieldsInRecordRange record : dataFile.records()) {
+	for (Record record : dataFile) {
 		ASSERT_LT(row, expectedFields.size()) << "Too many records";
 		unsigned col = 0;
-		for (std::string_view field : record) {
+		for (Field field : record) {
 			if (col < expectedFields[row].size())
-				EXPECT_EQ(field, expectedFields[row][col]) << "Unexpected value at record " << row << " and field " << col;
+				EXPECT_EQ(*field, expectedFields[row][col]) << "Unexpected value at record " << row << " and field " << col;
+			else
+				ADD_FAILURE() << "Extra value '" << field << "' in record " << row << " at field " << col;
+			col++;
+		}
+		EXPECT_GE(col, expectedFields[row].size()) << "Parsing returned fewer fields than expected in record " << row;
+		row++;
+	}
+	EXPECT_EQ(row, expectedFields.size()) << "Parsing returned fewer records than expected";
+}
+
+TEST(DataFileTest, ParseHeaderThenIterateOverRecords)
+{
+	auto result = LoadDataFile("txtdata\\lf.tsv");
+	ASSERT_TRUE(result.has_value()) << "Unable to load lf.tsv";
+
+	DataFile &dataFile = result.value();
+
+	std::vector<std::vector<std::string_view>> expectedFields {
+		{ "", "2", "3" },
+		{ "1", "2", "" },
+		{ "1", "", "3" },
+		{ "" } // file ends with a newline, parsing returns a single empty field

```

Comment:
Text files generally end with a newline, how come it is parsed as an empty field?

Shouldn't the following:

```
"a\nb\n"
```

Parse as: `{ "a" }, { "b" }`?

Reply:
That was initially to remove ambiguity, given inconsistencies between the way various programs produce tab-delimited files e.g. google sheets doesn't include a trailing newline on tsv exports). It seemed easiest to use newlines as separators at the time of #6399.

It's relatively trivial to change and I can make the behaviour less surprising by documenting the format as a collection of records terminated (not separated) by `\n` (with the last record having an optional terminator). Means I can get rid of the uses of the `Field::endOfFile` method too.